[TEP-0050] Implement PipelineTask OnError

QuanZhang-William opened the pull request:
<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

# Changes

Part of [#7165][#7165]. In [TEP-0050][tep-0050], we proposed to add an `OnError` API field under `PipelineTask` to configure error handling strategy.

This commits leverages `PipelineTask.OnError` API field introduced in the previous PR, implement the error handling strategy, update related docs and tests.

/kind feature

[tep-0050]: https://github.com/tektoncd/community/blob/main/teps/0050-ignore-task-failures.md
[#7165]: https://github.com/tektoncd/pipeline/issues/7165

<!-- Describe your changes here- ideally you can get that description straight from your descriptive commit message(s)! -->

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [ ] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) if any changes are user facing, including updates to minimum requirements e.g. Kubernetes version bumps
- [ ] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [ ] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [ ] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [ ] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [ ] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings). See some examples of [good release notes](https://github.com/tektoncd/community/blob/main/standards.md#good-release-notes).
- [ ] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
Implement "Ignore Task Failure" with new "PipelineTask.OnError" API field (TEP-0050). User can now set `pipelineTask.onError: continue` to ignore failure
```

QuanZhang-William commented:
/test all

dibyom commented on pkg/apis/pipeline/v1/pipelinerun_types.go:
```diff
@@ -412,6 +412,9 @@ const (
 	PipelineRunReasonInvalidParamValue PipelineRunReason = "InvalidParamValue"
 )
 
+// PipelineTaskOnErrorAnnotation is used to pass the failure strategy to TaskRun pods
```
nit: probably worth mentioning here that the failure strategy is from the pipelineTask's onError field

QuanZhang-William replied:
SG! I have updated the commnet, PTAL

QuanZhang-William commented on pkg/pod/status.go:
```diff
@@ -128,7 +128,12 @@ func MakeTaskRunStatus(ctx context.Context, logger *zap.SugaredLogger, tr v1.Tas
-		updateCompletedTaskRunStatus(logger, trs, pod)
+		onError, ok := tr.Annotations[v1.PipelineTaskOnErrorAnnotation]
+		if ok {
+			updateCompletedTaskRunStatus(logger, trs, pod, v1.PipelineTaskOnErrorType(onError))
```
Hi @JeromeJu. This is a good questions.

TEP-0050 is designed with requirement that `the task would be considered "successful" for the purposes of determining the status of the pipeline` (i.e. the `TaskRun` itself should still be failed, but with reason `FailureIgnored`). Looking at the [usecases](https://github.com/tektoncd/community/blob/main/teps/0050-ignore-task-failures.md#use-cases) in TEP-0050, it is mainly used to ignore pod error itself. 

I think your question is that for the validation/reconciler errors (like the `pvc creation`), should we also mark the reason as `FailureIgnored`? 

I think it might **not** be a good idea make the validation/reconciler errors  with reason `FailureIgnored`:
  - The validation/reconciler errors is likely from the Pipeline/Task Author, it shouldn't be ignored, but should be raised and fixed, `FailureIgnored` error reason could be confusing.
  - the error reason `FailureIgnored` becomes too generic if we use it across the TaskRun reconciler code. Instead, I'd prefer to keep the scope of `FailureIgnored` reason to the pod execution time.

WDYT?

JeromeJu replied:
Thanks @QuanZhang-William for the offline conversation and completing the 2nd half of my question!

And yes, I do agree that in actual use cases, only the container(Step) failures in the pod, which indicates users' own test failures, shall be ignored but not the "preparation" of those steps i.e. pvc creation and validations. Thanks for the explanation here.

So I think the current implementation is the way to go. One small ask is maybe we would can add this explanation/conversation somewhere in the docString or maybe link to the TEP050 for this design choice.

dibyom replied:
This makes sense as long as we document that `FailedIgnored` is ignoring a failure when the actual task execution fails but not otherwise (validation failure, etc.)

QuanZhang-William replied:
SG. I have added the docstring. PTAL

QuanZhang-William commented:
cc @tektoncd/core-collaborators @tektoncd/core-maintainers @jerop @pritidesai . Please take a look if you are interested!

dibyom commented:
Overall, this looks good to me! Thanks @QuanZhang-William Once you address the remaining open comments, I can approve.

QuanZhang-William commented on test/ignore_task_error_test.go:
```diff
@@ -0,0 +1,127 @@
+		t.Errorf("Expect failed-ignored-task Task in Failed status with reason %s, but got %s status with reason: %s", v1.TaskRunReasonFailureIgnored, cond.Status, cond.Reason)
+	}
+
+	// validate second TaskRun successed
```
Fixed, thanks!

QuanZhang-William commented:
> Overall, this looks good to me! Thanks @QuanZhang-William Once you address the remaining open comments, I can approve.

Thanks @dibyom. I've addressed the comments. PTAL!

JeromeJu commented on pkg/reconciler/pipelinerun/resources/pipelinerunstate.go:
```diff
@@ -278,11 +280,11 @@ func (state PipelineRunState) getNextTasks(candidateTasks sets.String) []*Resolv
 
 // IsStopping returns true if the PipelineRun won't be scheduling any new Task because
-// at least one task already failed or was cancelled in the specified dag
+// at least one task already failed (with onError: stopAndFail) or was cancelled in the specified dag
```
nit: "stopAndFail" is the default here this sounds like users have to set OnError?
Will leave this up to your decision.

JeromeJu reviewed (approved):
Thanks @QuanZhang-William 
One nit on release note, might be helpful to also include how users could use with the OnError feature i.e. "users can now set OnError to 'continue' to ignore failure from previous task".

dibyom reviewed (approved):
Any other feedback on this before merge? cc @vdemeester @jerop @pritidesai @afrittoli

QuanZhang-William commented:
> Any other feedback on this before merge? cc @vdemeester @jerop @pritidesai @afrittoli

Hi @dibyom. This PR has been opened for more than a week. Was wondering if we can merge? 🙏

Yongxuanzhang reviewed (commented):
/lgtm
